handbrake: fix Darwin build

bdesham opened the pull request:
###### Motivation for this change

The handbrake build had broken under Darwin. This PR fixes it by

- Omitting the numactl dependency under Darwin.
- Preventing the configure script from checking for xcodebuild. This check always failed (because xcodebuild isn't available in the build context) but xcodebuild isn't actually needed for the build, which uses `--disable-xcode`.
- Using the Nix-provided libxml2 under Darwin like we already do under Linux.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc @Anton-Latukha @wmertens

Anton-Latukha commented:
Interesting.

I previously talked to upstream developers about that `xcode` check, I even opened a report for it, they said that that check is optional.
Well, it is at least good to disable the check if it is not needed.

Does the present in file:
```
if stdenv.isDarwin then "--disable-xcode" else ""
```

works, or manual patching is indeed needed?

We can add `--force` into it if we really want.

Anton-Latukha commented:
It is great that you commented on the patches.

And it is also good to keep packages without manual regex patches.

As it frequently happens, the are some flags that upstream forgot to put in `./configure --help`, but they are in `make/configure.py`.

If the ` "--force --disable-xcode"` does not work - then we should remove it and patch.

Anton-Latukha commented:
I am OK with all the other changes.

Thank you for them.

It is just good for Nixpkgs to avoid patching when sources can change down the line.

bdesham commented:
@Anton-Latukha I tried adding `--force` to the configure options when building under Darwin, but unfortunately the configure script still stopped when it couldn’t find xcodebuild.

Anton-Latukha commented:
Ok. Thank you.

I approve the pull request.

Please, as you can - open a proper upstream bug report with according technical information.

bdesham commented on pkgs/applications/video/handbrake/default.nix:
```diff
@@ -84,6 +92,11 @@ _EOF
+    # Prevent the configure script from failing on Darwin if xcodebuild isn't
+    # available, which it isn't in the Nix context. (The actual build goes fine
+    # without xcodebuild.)
+    sed -e '/xcodebuild = ToolProbe/s/abort=.\+)/abort=False)/' -i make/configure.py
```
I will open a bug report with upstream when I have a chance.

bdesham replied:
It’s a little late but I opened https://github.com/HandBrake/HandBrake/issues/3048.

Anton-Latukha replied:
It is way better than never.

We can not permit the cost to gather all types of upstream patching, FP Nixpkgs is upstream Report generator, or otherwise, Nixpkgs would become just one huge patch of the IT world.

Anton-Latukha commented:
`HandBrake 1.3.3` is out, I would submit PR `1.3.(2->3)`, there is only ver/hash change, so PRs are independently mergeable. The release has minor changes, there are no changes related to this PR.

Anton-Latukha commented:
The reason there is no report about the issue the PR addresses - probably because their CI can just fail "optional" build, and they provide a set of prebuild packages, and for macOS also, we may be the only one package managemnet system that assembles and compiles the project on macOS, if in 8 month nobody except us noticed.

bdesham commented:
@Anton-Latukha It’s difficult for me to test a separate PR for the version bump because the package doesn’t build at all for me (on Darwin) without the changes in this current PR.

Anton-Latukha commented:
No, man, I just notified you that I opened parallel PR, and that they can merge in parallel.

Because I am quite aware of how parallel PRs can create all sorts of stuff. and so I inform you that there is parallel PR, and they would merge, no rebases needed.

I wish I could merge your PR myself.

bdesham commented:
Oh, I’m sorry, I misunderstood. I thought you were saying that *I* should open a new PR 🙂

Maybe I will mention this PR and #90302 on the NixOS Discourse to get some more visibility.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/170

Anton-Latukha commented:
I hope your PR gets merged faster. You somehow receive a huge timeouts on PRs merges.

wmertens commented on pkgs/applications/video/handbrake/default.nix:
```diff
@@ -76,6 +77,13 @@ _EOF
 
     patchShebangs scripts
 
+    # On Darwin, use the Nix-provided libxml2 instead of the patched version
```
these Darwin/Linux comments are confusing, neither line looks at the platform, so other platforms presumably get the Nix libxml2 as well?

bdesham replied:
If you look at [the file that’s being patched](https://github.com/HandBrake/HandBrake/blob/master/libhb/module.defs#L48-L64) you see that the patch labeled as Darwin here only applies to Darwin, Cygwin, and MinGW, and since Nix only supports Darwin out of those three, the patch only applies to Darwin. Similarly, the patch labeled as being for Linux is only applied on systems that are not FreeBSD, NetBSD, Darwin, Cygwin, or MinGW, which I believe means that it will only be applied under Linux.

With that being said, though, I agree that this is not really clear just from reading the derivation file. Since it could get messed up in the future if the upstream changes how things work on different platforms, it would be best to change the derivation so that it only makes certain patches when building for certain OSes.

After the discussion in HandBrake/HandBrake#3048, though, I think we should be using the HandBrake-patched versions of libxml2. The HandBrake developers do make an effort to push their changes upstream, and while we can debate whether shipping vendored dependencies is a good idea, I think it’s definitely not a good idea to swap one of HandBrake’s dependencies for a different version and then expect HandBrake to continue to work correctly. When I have some time I’ll look into getting the HandBrake derivation to use the HandBrake-patched version of libxml2.

wmertens replied:
Agreed, it's my fault, sorry. This was the first thing I did for nixpkgs and I split handbrake into the component packages because I thought it was cleaner. In reality many of those libraries are abandoned and applying the handbrake patches might not necessarily be a good idea.

Anton-Latukha replied:
Since Handbrake does a lot of package patching.
And since we can not impose those patches on the Nixpkgs, but they are expected in Handbrake.

It seems the solution is for Handbrake to have an overlay/override set for its dependencies.

wmertens replied:
not necessarily - it's been a while, but I believe it just keeps its own dependencies in its own tree and it builds them too. Adding an overlay seems like a lot of extra work.

Anton-Latukha replied:
> it just keeps its own dependencies in its own tree
Yep.
> it builds them too
Nope, we hack it to and force to use Nixpkgs dependencies:
https://github.com/NixOS/nixpkgs/blob/35521e4ea7a5bb471ce6b7286f5ce951b13a1027/pkgs/applications/video/handbrake/default.nix#L82-L83
dynamic deps is a policy of Nixpkgs.

Yes, Handbrake tries to build its deps shop by default, but it is in our shop )

What to do further is the question.

wmertens replied:
Ah - originally I did break it out but then we agreed not to do that after all https://github.com/NixOS/nixpkgs/pull/2482#issuecomment-42113475

bdesham replied:
I modified the patching so that it’s more obvious which things are being patched on which platforms. For the time being I’m leaving things as they are with Handbrake’s dependencies: we use the Nixpkgs-provided versions, not the Handbrake patched versions.

wmertens reviewed (commented):
LGTM, just need some more clarity about those libxml2 patches?

Anton-Latukha commented:
@danieldk

I see you merged active recently.

Two package maintainers reviewed, it, changes were discussed and made, and PR approved by two maintainer. We wait for someone to come around and merge this for some time.

Pinging, we close the https://github.com/NixOS/nixpkgs/issues/99317 report.

veprbl commented on pkgs/applications/video/handbrake/default.nix:
```diff
@@ -10,7 +10,7 @@
   # Main build tools
   pkgconfig, autoconf, automake, libtool, m4, lzma, python3,
-  numactl,
+  numactl ? null,
```
This serves no purpose as `numactl` is always defined: https://github.com/NixOS/nixpkgs/blob/e6ec27be9808ae6e661e500e9bc31f959ac449bc/pkgs/top-level/all-packages.nix#L18268

veprbl commented on pkgs/applications/video/handbrake/default.nix:
```diff
@@ -48,6 +48,7 @@ assert stdenv.isDarwin -> AudioToolbox != null
        && Foundation != null
        && libobjc != null
        && VideoToolbox != null;
+assert !stdenv.isDarwin -> numactl != null;
```
In my opinion, asserts like this don't prevent any mistake one could conceivably make, but may cause unnecessary headache when doing overrides (see #53487 for details)

bdesham replied:
Is it your opinion that the statement before this one is also superfluous?

veprbl replied:
@bdesham Well, you are the author of 765cd8c2edb8c3406b49f772bb37eaa271c4b2b2, so you might know why it is needed. After a quick check I still don't see a practical use case.

bdesham replied:
This has been rebased and updated to exclude those `assert`s.

veprbl replied:
Thanks! Let's mark this as resolved then. If you come up with a reason why the asserts are useful, I would be curious to know, feel free to ping me.

bdesham commented:
@GrahamcOfBorg build handbrake

veprbl commented on pkgs/applications/video/handbrake/default.nix:
```diff
@@ -76,15 +71,26 @@ _EOF
         -e '/    ## Additional library and tool checks/,/    ## MinGW specific library and tool checks/d' \
         -i make/configure.py
-  '';
+  '' + (if stdenv.isDarwin then ''
```
```suggestion
  '' + (lib.optionalString stdenv.isDarwin ''
```

bdesham replied:
Thanks, I thought there was probably a more elegant way to do that.

veprbl reviewed (approved):
LGTM, runs on x86_64-darwin

bdesham committed:
handbrake: fix Darwin build

- Omit the numactl dependency under Darwin.
- Prevent the configure script from checking for xcodebuild. This check
  always failed (because xcodebuild isn't available in the build
  context) but xcodebuild isn't actually needed for the build, which
  uses --disable-xcode.
- Use the Nix-provided libxml2 under Darwin like we already do under
  Linux.

bdesham commented:
@GrahamcOfBorg build handbrake

bdesham commented:
This is (once again) ready to be merged.

veprbl commented:
Thank you!
